### PR TITLE
Remove legacy storage

### PR DESCRIPTION
**Pull request by wolverineks**

### CODE REVIEW

File: src/legacy/DeveloperScreen.tsx [hunk 1]
```diff
@@ -204,3 +207,87 @@ export const DeveloperScreen = () => {
     </SafeAreaView>
   )
 }
+
```
Comment (wolverineks) [inline src/legacy/DeveloperScreen.tsx:210]: copied here, too lazy to migrate this component

File: src/legacy/actions.ts [hunk 2]
```diff
@@ -7,21 +7,22 @@ import uuid from 'uuid'
...
 import {walletManager} from '../yoroi-wallets'
+import {YoroiStorage} from '../yoroi-wallets/storage'
+import {parseString} from '../yoroi-wallets/utils/parsing'
 import assert from './assert'
 import crashReporting from './crashReporting'
 
-const initInstallationId = async (storage: Storage) => {
-  const installationId = await storage.getItem('/appSettings/installationId')
+const initInstallationId = async (storage: YoroiStorage) => {
+  const installationId = await storage.join('appSettings/').getItem('installationId', parseString)
```
Comment (wolverineks) [inline src/legacy/actions.ts]: double check this

File: src/hooks/index.ts [hunk 3]
```diff
@@ -179,10 +180,11 @@ export const useCloseWallet = (options: UseMutationOptions<void, Error> = {}) =>
 }
 
 export const useWalletName = (wallet: YoroiWallet, options?: UseQueryOptions<string, Error>) => {
+  const storage = useStorage()
```
Comment (wolverineks) [inline src/hooks/index.ts:183]: eventually migrate to wallet.storage

File: src/hooks/index.ts [hunk 4]
```diff
@@ -194,12 +196,13 @@ export const useWalletName = (wallet: YoroiWallet, options?: UseQueryOptions<str
 }
 
 export const useChangeWalletName = (wallet: YoroiWallet, options: UseMutationOptions<void, Error, string> = {}) => {
+  const storage = useStorage()
```
Comment (wolverineks) [inline src/hooks/index.ts:199]: eventually, migrate to wallet.storage

File: src/index.tsx [hunk 5]
```diff
@@ -61,26 +62,26 @@ global.onunhandledrejection = (error: any) => {
...
   return migrated ? (
-    <WalletManagerProvider walletManager={walletManager}>
-      <ErrorBoundary>
-        <QueryClientProvider client={queryClient}>
-          <LoadingBoundary style={StyleSheet.absoluteFill}>
-            <ThemeProvider>
-              <LanguageProvider>
-                <CurrencyProvider>
-                  <StorageProvider>
+    <StorageProvider>
```
Comment (wolverineks) [inline src/index.tsx:68]: move StorageProvider to outer most provider

File: src/legacy/actions.ts [hunk 6]
```diff
@@ -3,35 +3,9 @@ import 'react-intl'
 
 import type {IntlShape} from 'react-intl'
 import {Alert} from 'react-native'
-import uuid from 'uuid'
```
Comment (wolverineks) [inline src/legacy/actions.ts:6]: moved to App.tsx

File: src/yoroi-wallets/cardano/ShelleyWallet.ts [hunk 7]
```diff
@@ -1079,9 +1079,8 @@ export class ShelleyWallet implements WalletInterface {
   async changePassword(oldPassword: string, newPassword: string) {
     if (!_.isEmpty(validatePassword(newPassword, newPassword))) throw new Error('New password is not valid')
 
-    const key = EncryptedStorageKeys.rootKey(this.id)
-    const rootKey = await EncryptedStorage.read(key, oldPassword)
-    return EncryptedStorage.write(key, rootKey, newPassword)
+    const rootKey = await this.encryptedStorage.rootKey.read(oldPassword)
```
Comment (wolverineks) [inline src/yoroi-wallets/cardano/ShelleyWallet.ts:1082]: migrate from module to instance storage

File: src/yoroi-wallets/storage.ts [hunk 8]
```diff
@@ -29,12 +29,12 @@ const mountStorage = (path: Path) => {
 
     getItem,
     multiGet,
-    setItem: (key: string, value: unknown) => {
-      const item = JSON.stringify(value)
+    setItem: <T = unknown>(key: string, value: T, stringify: (data: T) => string = JSON.stringify) => {
```
Comment (wolverineks) [inline src/yoroi-wallets/storage.ts:32]: enable custom stringify

File: src/yoroi-wallets/storage.ts [hunk 9]
```diff
@@ -29,12 +29,12 @@ const mountStorage = (path: Path) => {
...
+      const item = stringify(value)
       return AsyncStorage.setItem(withPath(key), item)
     },
-    multiSet: (tuples: Array<[key: string, value: unknown]>) => {
-      const items = tuples.map(([key, value]) => [withPath(key), JSON.stringify(value)])
+    multiSet: (tuples: Array<[key: string, value: unknown]>, stringify: (data: unknown) => string = JSON.stringify) => {
```
Comment (wolverineks) [inline src/yoroi-wallets/storage.ts:36]: enable custom stringify

File: src/auth/hooks.ts [hunk 10]
```diff
@@ -38,27 +39,29 @@ export const useAuthOsEnabled = (options?: UseQueryOptions<boolean, Error>) => {
...
     },
-    {authenticationPrompt},
-  )
+  })
 
-  return {...mutation, enableAuthWithOs}
+  return {...mutation, enableAuthWithOs: enable}
+}
+
+export const enableAuthWithOs = async (storage: YoroiStorage) => {
```
Comment (wolverineks) [inline src/auth/hooks.ts:57]: extract for testing

File: src/auth/hooks.ts [hunk 11]
```diff
@@ -177,48 +181,35 @@ export const useDisableAllEasyConfirmation = (
...
+export const disableAllEasyConfirmation = async (storage: YoroiStorage) => {
+  const walletIds = await storage.join('wallet/').getAllKeys()
+
+  const updateWalletMetas = walletIds.map(async (walletId) => {
+    const walletMeta = await storage.join('wallet/').getItem(walletId, parseWalletMeta)
+    await storage.join('wallet/').setItem(walletId, {...walletMeta, isEasyConfirmationEnabled: false})
+  })
+
+  const updateWalletJSONs = walletIds.map(async (walletId) => {
+    const walletJSON: WalletJSON = await storage.join('wallet/').join(`${walletId}/`).getItem('data')
```
Comment (wolverineks) [inline src/auth/hooks.ts]: assumes valid `WalletJSON`

File: src/auth/hooks.ts [hunk 12]
```diff
@@ -268,22 +258,18 @@ export const useAuthSetting = (storage: Storage, options?: UseQueryOptions<AuthS
...
 
 const parseAuthSetting = (data: unknown) => {
-  if (!data) return undefined
-  try {
-    return JSON.parse(data as string)
-  } catch (error) {
-    return undefined
-  }
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  const isAuthSetting = (data: any): data is 'os' | 'pin' | undefined => ['os', 'pin', undefined].includes(data)
```
Comment (wolverineks) [inline src/auth/hooks.ts:265]: why is `undefined` a valid auth setting?

File: src/auth/hooks.ts [hunk 13]
```diff
@@ -268,22 +258,18 @@ export const useAuthSetting = (storage: Storage, options?: UseQueryOptions<AuthS
   return query.data
 }
 
-export const getAuthSetting = async (storage: Storage) => {
-  const authSetting = parseAuthSetting(await storage.getItem(SettingsStorageKeys.Auth))
-  if (isAuthSetting(authSetting)) return authSetting
+export const getAuthSetting = async (storage: YoroiStorage) => {
+  const authSetting = await storage.join('appSettings/').getItem('auth', parseAuthSetting)
+  if (authSetting !== null) return authSetting
   return Promise.reject(new Error('useAuthSetting invalid data'))
```
Comment (wolverineks) [inline src/auth/hooks.ts]: how does the app handle invalid auth setting?
is there a sane default value?

File: src/auth/hooks.ts (hunk 12 above)
Comment (stackchain) [reply src/auth/hooks.ts:265]: during the first run / migration, it is, there is no default value

File: src/auth/hooks.ts (hunk 13 above)
Comment (stackchain) [reply src/auth/hooks.ts]: unreachable code at 264, it fallbacks to `undefined`, maybe this helper doesn't make sense anymore and can be moved within the query,  `() => storage.join('appSettings/').getItem('auth', parseAuthSetting)`

File: src/yoroi-wallets/storage.ts [hunk 14]
```diff
@@ -2,7 +2,7 @@ import AsyncStorage from '@react-native-async-storage/async-storage'
 
 import {parseSafe} from './utils/parsing'
 
-export type Storage = ReturnType<typeof mountStorage>
+export type YoroiStorage = ReturnType<typeof mountStorage>
 export type Path = `${string}/`
 
 const mountStorage = (path: Path) => {
```
Comment (stackchain) [inline src/yoroi-wallets/storage.ts:8]: return as const

File: src/Settings/PrivacyMode/PrivacyMode.tsx [hunk 15]
```diff
@@ -1,14 +1,16 @@
...
 export const usePrivacyMode = () => {
+  const storage = useStorage()
   const query = useQuery<PrivacyMode, Error>({
     queryKey: ['privacyMode'],
     queryFn: async () => {
-      const storedPrivacyMode = await AsyncStorage.getItem('/appSettings/privacyMode')
+      const storedPrivacyMode = await storage.join('appSettings/').getItem('privacyMode', parseString)
 
       if (!isEmptyString(storedPrivacyMode)) {
         const parsedPrivacyMode = JSON.parse(storedPrivacyMode)
```
Comment (wolverineks) [inline src/Settings/PrivacyMode/PrivacyMode.tsx]: fix this

File: src/Storage/migrations/4_9_0.ts [hunk 16]
```diff
@@ -1,25 +1,31 @@
...
 import {parseBoolean} from '../../yoroi-wallets/utils/parsing'
-import {SettingsStorageKeys, Storage} from '../StorageProvider'
 
-export const migrateAuthSetting = async (storage: Storage) => {
+export const migrateAuthSetting = async (storage: YoroiStorage) => {
   const authSetting = await getAuthSetting(storage)
-  const isFirstRun = await storage.getItem(SettingsStorageKeys.InstallationId).then((data) => data === null)
+  const isFirstRun = await storage
+    .join('appSettings/')
+    .getItem('installationId')
```
Comment (wolverineks) [inline src/Storage/migrations/4_9_0.ts:9]: parseInstallationId

Comment (wolverineks) [reply src/Storage/migrations/4_9_0.ts:9]: data -> value

File: src/Storage/migrations/4_9_0.ts [hunk 17]
```diff
@@ -1,25 +1,31 @@
...
+    .then((data) => data === null)
```
Comment (stackchain) [inline src/Storage/migrations/4_9_0.ts]:
```suggestion
    .then((value) => value === null)
```

**Review by stackchain — APPROVED** [review]

**Pull request merged**